src/Catalog/PackageCatalogItem.cs

```diff
@@ -130,31 +134,54 @@ public override IGraph CreateContentGraph(CatalogContext context)
 
                 if (Deprecation.AlternatePackageId != null)
                 {
-                    // assert alternate package root node to deprecation root node
+                    //  assert alternate package root node to deprecation root node
                     var deprecationAlternatePackagePredicate = graph.CreateUriNode(Schema.Predicates.AlternatePackage);
                     var deprecationAlternatePackageRootNode = graph.CreateUriNode(new Uri(resource.Subject.ToString() + "#deprecation/alternatePackage"));
                     graph.Assert(deprecationRootNode, deprecationAlternatePackagePredicate, deprecationAlternatePackageRootNode);
 
-                    // assert id to alternate package root node
+                    //  assert id to alternate package root node
                     graph.Assert(
                         deprecationAlternatePackageRootNode,
                         graph.CreateUriNode(Schema.Predicates.Id),
                         graph.CreateLiteralNode(Deprecation.AlternatePackageId));
 
-                    // assert version range to alternate package root node
+                    //  assert version range to alternate package root node
                     graph.Assert(
                         deprecationAlternatePackageRootNode,
                         graph.CreateUriNode(Schema.Predicates.Range),
                         graph.CreateLiteralNode(Deprecation.AlternatePackageRange));
                 }
             }
 
+            //  vulnerabilities
+            if (Vulnerabilities != null)
+            {
+                INode vulnerabilityPredicate = graph.CreateUriNode(Schema.Predicates.Vulnerability);
+                INode vulnerabilityType = graph.CreateUriNode(Schema.DataTypes.Vulnerability);
+                INode gitHubDatabaseKeyPredicate = graph.CreateUriNode(Schema.Predicates.GitHubDatabaseKey);
+                INode advisoryUrlPredicate = graph.CreateUriNode(Schema.Predicates.AdvisoryUrl);
+                INode severityPredicate = graph.CreateUriNode(Schema.Predicates.Severity);
+
+                foreach (PackageVulnerabilityItem vulnerability in Vulnerabilities)
+                {
+                    Uri vulnerabilityUri = new Uri(resource.Subject.ToString() + "#" + vulnerability.GitHubDatabaseKey);
```

Comment:
What is the type of `Subject`? `Uri`? If so `ToString()` should not be used. Instead, `AbsoluteUri` should be used. Alternatively, you can use `UriBuilder` to ensure we don't concat a fragment after another fragment.

Reply:
See https://github.com/NuGet/NuGet.Services.Metadata/blob/e4b4b1392492a04a9e3b0cf72bb04a303338ede1/src/Catalog/PackageCatalogItem.cs#L89. I can change both to `AbsoluteUri`, but if the resultant string from both approaches is different, could we possibly pollute data in the PackageEntry collections by changing this now?

Reply:
Given the current pattern of URLs the behavior of `ToString()` should match `AbsoluteUri` in the vast majority of cases. Suspicious cases are when the ID has non-ASCII or non-URL safe characters. There are known problems with `Uri.ToString()` such as decoding URL characters for display purposes but incorrectly for data purposes. See this blog for more details: https://dhvik.blogspot.com/2019/12/uritostring-automatically-decodes-url.html

There is no known consumer of package entries so the risk is very low. If you are concerned have the bandwidth, sample a couple of cases to see if anything jumps out.

If you want to avoid the analysis, you can make yours correct and file a bug on the other.

Reply:
@joelverhagen Did a little research, and the type is `VDS.RDF.UriNode`, not `Uri`, and its `ToString()` method does indeed return `AbsoluteUri`, as we can see in its base class here:
https://github.com/dotnetrdf/dotnetrdf/blob/1e3a6228eea930bb2d7a2b0fb2ce47eadb71d4a5/Libraries/dotNetRDF/Core/URINode.cs#L262
So we're good to go on both of these counts.

Reply:
Awesome, thanks for investigating this!